scantool/diag_general.c

```diff
@@ -323,52 +326,40 @@ diag_geterr(void) {
 
 /* Memory allocation */
 
-//diag_flcalloc (srcfilename, srcfileline, ptr, num,size) = allocate (num*size) bytes
-// also checks for size !=0
-// ret 0 if ok
-int diag_flcalloc(const char *name, const int line,
-	void **pp, size_t n, size_t s) {
-	void *p;
-
-	//sanity check: make sure we weren't given a null ptr
-	//or a null size.
+// Stores pointer to a newly allocated buffer of n*s bytes to pp.
+// Also takes filename and line to report for debugging purposes.
+// Returns 0 in the absence of errors.
+int diag_fl_alloc(const char *fName, const int line,
+	void **pp, size_t n, size_t s, bool allocIsCalloc) {
+	char allocator;
+	char *errMsg = "%s:%d: %calloc(%ld, %ld) failed: %s\n";
 
-	if ((s !=0) && (pp != NULL)) {
-		p = calloc(n, s);
+	if (allocIsCalloc) {
+		allocator = 'c';
 	} else {
-		p=NULL;
+		allocator = 'm';
 	}
 
-	*pp = p;
-
-	if (p == NULL) {
-		fprintf(stderr,
-			"%s:%d: calloc(%ld, %ld) failed: %s\n", name, line,
-			(long)n, (long)s, strerror(errno));
-		return diag_iseterr(DIAG_ERR_NOMEM);
+	// Check for null or overflow-causing parameters.
+	if (pp == NULL || n == 0 || s == 0 || SIZE_MAX/s < n) {
```

Comment:
I'd like to see some more spaces and/or parens in that overflow check (it can still stay on a single line), at least to clear up that "SIZE_MAX/s"

Reply:
Done.